File: data/snippets/quiz020.cpp
class Address {
private:
    std::string street;
    std::string city;
    std::string postalCode;
    std::string country;
    
public:
    Address(std::string str, std::string c, std::string pc, std::string cntry) 
        : street(str), city(c), postalCode(pc), country(cntry) {}
        
    std::string getStreet() const { return street; }
    std::string getCity() const { return city; }
    std::string getPostalCode() const { return postalCode; }
    std::string getCountry() const { return country; }
};

class Supplier {
private:
    std::string name;
    Address address;
    ContactInfo contactInfo;
    
public:
    Supplier(std::string n, Address a, ContactInfo ci) 
        : name(n), address(a), contactInfo(ci) {}
        
    std::string getName() const { return name; }
    Address getAddress() const { return address; }
    ContactInfo getContactInfo() const { return contactInfo; }
};

class ContactInfo {
private:
    std::string email;
    PhoneNumber phoneNumber;
    std::string website;
    
public:
    ContactInfo(std::string e, PhoneNumber pn, std::string web) 
        : email(e), phoneNumber(pn), website(web) {}
        
    std::string getEmail() const { return email; }
    PhoneNumber getPhoneNumber() const { return phoneNumber; }
    std::string getWebsite() const { return website; }
};

class PhoneNumber {
private:
    std::string countryCode;
    std::string areaCode;
    std::string number;
    
public:
    PhoneNumber(std::string cc, std::string ac, std::string num) 
        : countryCode(cc), areaCode(ac), number(num) {}
        
    std::string getCountryCode() const { return countryCode; }
    std::string getAreaCode() const { return areaCode; }
    std::string getNumber() const { return number; }
    
    std::string getFormattedNumber() const {
        return "+" + countryCode + " " + areaCode + " " + number;
    }
};

class Product {
private:
    std::string name;
    double price;
    Supplier supplier;
    
public:
    Product(std::string n, double p, Supplier s) 
        : name(n), price(p), supplier(s) {}
        
    std::string getName() const { return name; }
    double getPrice() const { return price; }
    Supplier getSupplier() const { return supplier; }
};

class Inventory {
private:
    std::vector<Product> products;
    
public:
    void addProduct(const Product& product) {
        products.push_back(product);
    }
    
    Product getProduct(int index) const {
        return products[index];
    }
    
    int getProductCount() const {
        return products.size();
    }
};

// 客户端代码使用长消息链
void clientCode(const Inventory& inventory) {
    // 获取产品0的供应商的联系信息的电话号码的格式化字符串
    std::string phoneNumber = inventory.getProduct(0).getSupplier().getContactInfo().getPhoneNumber().getFormattedNumber();
    std::cout << "Supplier phone: " << phoneNumber << std::endl;
    
    // 获取产品0的供应商的地址的城市
    std::string city = inventory.getProduct(0).getSupplier().getAddress().getCity();
    std::cout << "Supplier city: " << city << std::endl;
    
    // 获取产品0的供应商的国家
    std::string country = inventory.getProduct(0).getSupplier().getAddress().getCountry();
    std::cout << "Supplier country: " << country << std::endl;
}

File: data/snippets/quiz020.java
import java.util.List;
import java.util.ArrayList;

class Address {
    private String street;
    private String city;
    private String postalCode;
    private String country;
    
    public Address(String street, String city, String postalCode, String country) {
        this.street = street;
        this.city = city;
        this.postalCode = postalCode;
        this.country = country;
    }
    
    public String getStreet() { return street; }
    public String getCity() { return city; }
    public String getPostalCode() { return postalCode; }
    public String getCountry() { return country; }
}

class PhoneNumber {
    private String countryCode;
    private String areaCode;
    private String number;
    
    public PhoneNumber(String countryCode, String areaCode, String number) {
        this.countryCode = countryCode;
        this.areaCode = areaCode;
        this.number = number;
    }
    
    public String getCountryCode() { return countryCode; }
    public String getAreaCode() { return areaCode; }
    public String getNumber() { return number; }
    
    public String getFormattedNumber() {
        return "+" + countryCode + " " + areaCode + " " + number;
    }
}

class ContactInfo {
    private String email;
    private PhoneNumber phoneNumber;
    private String website;
    
    public ContactInfo(String email, PhoneNumber phoneNumber, String website) {
        this.email = email;
        this.phoneNumber = phoneNumber;
        this.website = website;
    }
    
    public String getEmail() { return email; }
    public PhoneNumber getPhoneNumber() { return phoneNumber; }
    public String getWebsite() { return website; }
}

class Supplier {
    private String name;
    private Address address;
    private ContactInfo contactInfo;
    
    public Supplier(String name, Address address, ContactInfo contactInfo) {
        this.name = name;
        this.address = address;
        this.contactInfo = contactInfo;
    }
    
    public String getName() { return name; }
    public Address getAddress() { return address; }
    public ContactInfo getContactInfo() { return contactInfo; }
}

class Product {
    private String name;
    private double price;
    private Supplier supplier;
    
    public Product(String name, double price, Supplier supplier) {
        this.name = name;
        this.price = price;
        this.supplier = supplier;
    }
    
    public String getName() { return name; }
    public double getPrice() { return price; }
    public Supplier getSupplier() { return supplier; }
}

class Inventory {
    private List<Product> products;
    
    public Inventory() {
        this.products = new ArrayList<>();
    }
    
    public void addProduct(Product product) {
        products.add(product);
    }
    
    public Product getProduct(int index) {
        return products.get(index);
    }
    
    public int getProductCount() {
        return products.size();
    }
}

class ClientCode {
    public void processInventory(Inventory inventory) {
        String phoneNumber = inventory.getProduct(0).getSupplier().getContactInfo().getPhoneNumber().getFormattedNumber();
        System.out.println("Supplier phone: " + phoneNumber);
        
        String city = inventory.getProduct(0).getSupplier().getAddress().getCity();
        System.out.println("Supplier city: " + city);
        
        String country = inventory.getProduct(0).getSupplier().getAddress().getCountry();
        System.out.println("Supplier country: " + country);
        
        String email = inventory.getProduct(0).getSupplier().getContactInfo().getEmail();
        System.out.println("Supplier email: " + email);
    }
}
